### ISSUE TITLE
Faster conversion to String

### ISSUE THREAD

**Issue by jwilm** [uuid-rs/uuid#196]
Hello, I had a need for much faster printing of UUIDs, and I created a library [uuidtoa](https://github.com/jwilm/uuidtoa) to prototype and use immediately in another project. I would like to send a PR to include the improved write speed here, but a couple of questions first.

1. Is the uuid project interested?
2. Not going through the format! machinery is a requirement for me. Would you be open to adding a public API like what's used in uuidtoa _in addition_ to the `fmt::Display` way of printing currently included in uuid?

Thanks!

**Comment by kinggoesgaming** [comment]
If we decide to go ahead with this... I would suggest waiting for about a month (currently working a whole set of breaking changes, and holding up on new features)

**Comment by KodrAus** [comment]
Thanks for building this @jwilm! I'm on board with pulling it in and improving our `Display` implementation as well as offering an alternative lightweight way to write out UUIDs. Maybe we could add a `fmt` module that includes the `write_*` methods?

**Comment by kinggoesgaming** [comment]
for the record me and @Dylan-DPC are onboard for this aswell... 

@jwilm mind joining our [gitter](https://gitter.im/uuid-rs/Lobby) to discuss this?

**Comment by Dylan-DPC** [comment]
@jwilm any updates on this?

**Comment by jwilm** [comment]
I won't personally have time to do this for a while, but you are welcome to pull in the code.

**Comment by Dylan-DPC** [comment]
Uhmm no issues. We still have time for the next major release (0.7)

**Comment by jwilm** [comment]
The format changes won't have breakage anyhow. The only API difference would be the _addition_ of a `fmt` module.

**Comment by kinggoesgaming** [comment]
At this point, we have frozen `0.6.x` cycle sans bug fixes... It's just keep the load light on us

**Comment by kinggoesgaming** [comment]
@huonw I think your PR fixes this right? we just don't have to have separate functions anymore for the similar speedups

**Comment by huonw** [comment]
It probably puts uuid similar to uuidtoa, but I suspect uuidtoa will still be slightly faster, and the uuid code still goes through the `fmt` machinery. Maybe @jwilm could compare the performance again.

**Comment by kinggoesgaming** [comment]
for reference: PR #303 

available in `uuid@0.7.0-beta` and onwards

**Comment by huonw** [comment]
I tried benchmarking it myself, using https://github.com/huonw/uuid/commit/e8504e4f5f6ccf606b9cb134431c1ed1dd625e84.

Results without LTO:

```txt
running 4 tests
test bench_hyphen  ... bench:          49 ns/iter (+/- 4)
test bench_simple  ... bench:          40 ns/iter (+/- 4)
test bench_urn     ... bench:          56 ns/iter (+/- 3)
test bench_uuidtoa ... bench:          66 ns/iter (+/- 7)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out
```

Results with LTO:

```txt
running 4 tests
test bench_hyphen  ... bench:          42 ns/iter (+/- 6)
test bench_simple  ... bench:          34 ns/iter (+/- 2)
test bench_urn     ... bench:          47 ns/iter (+/- 8)
test bench_uuidtoa ... bench:          14 ns/iter (+/- 1)

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out
```

That is, in the full-optimisations case, there's ~30ns of overhead using `uuid` directly (maybe 5ns of this is due to using `Formatter::write_str` and thus calling `str::from_utf8` instead of `str::from_utf8_unchecked`).

---

One possibility for an API might be something like along the lines of https://doc.rust-lang.org/std/primitive.char.html#method.encode_utf8 e.g.

```rust
impl Hyphenated {
    fn encode_lower(&self, v: &mut [u8]) -> &mut str { ... }
    fn encode_upper(&self, v: &mut [u8]) -> &mut str { ... }
}
```

where the `v` must have length at least 36 (or 32, or 45, depending on the adapter). This can then be (somewhat) fluently used for writing to an intermediate buffer:

```rust
let mut buf: &mut [u8] = ...;

let written = uuid.to_hyphenated().encode_lower(&mut buf).len();
buf = &mut buf[written..];
```

And it can also be used naturally if the `str` encoded value is all that's useful:

```rust
fn foo(s: &str) { ... }
foo(uuid.to_hyphenated().encode_lower(&mut [0; 36]))
```

This also avoids double buffering: the function doesn't need to create its own `[u8; N]` buffer internally that it then copies into the real output.

Variations/additions:

- take a flag for upper/lower as a parameter instead of two separate functions
- provide a function for creating the buffer, e.g. `Uuid::encode_buffer() -> [u8; 45]`, so the latter case can be `foo(uuid.to_hyphenated().encode_lower(&mut Uuid::encode_buffer()))` without the programmer having think about the size of a printed UUID
- offer `encode_...` on `Uuid` itself, probably using the hyphenated form (this would probably best be in addition to the others, meaning the common case is `uuid.encode_lower(&mut [0; 36])`, but the simple and urn forms are still easily accessible).

Thoughts?
